nav2_map_server/src/vo_server/vector_object_server.cpp

```diff
@@ -0,0 +1,551 @@
+  } else if (size_x > 0 && size_y > 0) {
+    // Map size was not changed
+    memset(map_->data.data(), default_value_, size_x * size_y * sizeof(int8_t));
+  }
+  map_->header.frame_id = global_frame_id_;
+  map_->info.resolution = resolution_;
+  map_->info.origin.position.x = min_x;
+  map_->info.origin.position.y = min_y;
+}
+
+void VectorObjectServer::putVectorObjectsOnMap()
+{
+  // Filling the shapes
+  for (auto shape : shapes_) {
+    if (shape.second->isFill()) {
+      // Put filled shape on map
+      double wx1 = std::numeric_limits<double>::max();
+      double wy1 = std::numeric_limits<double>::max();
+      double wx2 = std::numeric_limits<double>::lowest();
+      double wy2 = std::numeric_limits<double>::lowest();
+      unsigned int mx1 = 0;
+      unsigned int my1 = 0;
+      unsigned int mx2 = 0;
+      unsigned int my2 = 0;
+
+      shape.second->getBoundaries(wx1, wy1, wx2, wy2);
+      if (
+        !nav2_util::worldToMap(map_, wx1, wy1, mx1, my1) ||
+        !nav2_util::worldToMap(map_, wx2, wy2, mx2, my2))
+      {
+        RCLCPP_ERROR(
+          get_logger(),
+          "Error to get shape boundaries on map (UUID: %s)",
+          shape.second->getUUID().c_str());
+        return;
+      }
+
+      unsigned int it;
+      for (unsigned int my = my1; my <= my2; my++) {
+        for (unsigned int mx = mx1; mx <= mx2; mx++) {
+          it = my * map_->info.width + mx;
+          double wx, wy;
+          nav2_util::mapToWorld(map_, mx, my, wx, wy);
+          if (shape.second->isPointInside(wx, wy)) {
+            processVal(map_->data[it], shape.second->getValue(), overlay_type_);
+          }
+        }
+      }
+    } else {
+      // Put shape borders on map
+      shape.second->putBorders(map_, overlay_type_);
+    }
+  }
+}
+
+void VectorObjectServer::publishMap()
+{
+  if (map_) {
+    auto map = std::make_unique<nav_msgs::msg::OccupancyGrid>(*map_);
+    map_pub_->publish(std::move(map));
+  }
+}
+
+void VectorObjectServer::processMap()
+{
+  if (!process_map_) {
+    return;
+  }
+
+  try {
+    if (shapes_.size() > 0) {
+      if (!transformVectorObjects()) {
+        return;
+      }
+      double min_x, min_y, max_x, max_y;
+
+      getMapBoundaries(min_x, min_y, max_x, max_y);
+      updateMap(min_x, min_y, max_x, max_y);
+      putVectorObjectsOnMap();
+    } else {
+      updateMap(0.0, 0.0, 0.0, 0.0);
+    }
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(get_logger(), "Can not upate map: %s", ex.what());
+    return;
+  }
+
+  publishMap();
+}
+
+void VectorObjectServer::switchMapUpdate()
+{
+  for (auto shape : shapes_) {
+    if (shape.second->getFrameID() != global_frame_id_ && !shape.second->getFrameID().empty()) {
+      if (!map_timer_) {
+        map_timer_ = this->create_timer(
+          std::chrono::duration<double>(1.0 / update_frequency_),
+          std::bind(&VectorObjectServer::processMap, this));
+      }
+      RCLCPP_INFO(get_logger(), "Publishing map dynamically");
```

Comment:
Update to mention what rate its publishing at, that's useful debug info

Reply:
Fixed in https://github.com/ros-planning/navigation2/pull/3930/commits/43af6df733237b6f5ef8c91d0757b2ceda43a2de